drivers/platform/stm32/stm32_xspi.c

```diff
@@ -0,0 +1,718 @@
+			xspi_cmd->IOSelect = 0;
+			xspi_cmd->AddressDTRMode = 0;
+			xspi_cmd->AlternateBytesDTRMode = 0;
+			xspi_cmd->DataDTRMode = 0;
+			xspi_cmd->DQSMode = 0;
+			xspi_cmd->InstructionDTRMode = HAL_XSPI_INSTRUCTION_DTR_DISABLE;
+			xspi_cmd->SIOOMode = 0;
+		}
+	}
+
+	/* Configure the XSPI command */
+	ret = HAL_XSPI_Command(&xspi_desc->hxspi, xspi_cmd,
+			       HAL_XSPI_TIMEOUT_DEFAULT_VALUE);
+	if (ret)
+		return ret;
+
+	/* Explicitly disable address mode */
+	OCTOSPI1->CCR &= ~OCTOSPI_CCR_ADMODE_Msk;
+	OCTOSPI1->CCR &= ~OCTOSPI_CCR_IMODE_Msk;
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the XSPI communication peripheral.
+ * @param desc - The XSPI descriptor.
+ * @param param - The structure that contains the XSPI parameters.
+ * @return 0 in case of success, negative code otherwise.
+ */
+int32_t stm32_xspi_init(struct no_os_xspi_desc **desc,
+			const struct no_os_xspi_init_param *param)
+{
+	int32_t ret;
+	struct no_os_xspi_desc *xspi_desc;
+	struct stm32_xspi_desc *sdesc;
+	struct stm32_xspi_init_param *sinit;
+	XSPI_RegularCmdTypeDef *xspi_cmd;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	if (!param->extra)
+		return -EINVAL;
+
+	sinit = (struct stm32_xspi_init_param *) param->extra;
+
+	/* Allocate memory for no-os xspi descriptor */
+	xspi_desc = no_os_calloc(1, sizeof(*xspi_desc));
+	if (!xspi_desc) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	/* Allocate memory for stm32 xspi descriptor */
+	sdesc = no_os_calloc(1, sizeof(*sdesc));
+	if (!sdesc) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	/* Allocate memory for STM32 specific platform XSPI command */
+	xspi_cmd = no_os_calloc(1, sizeof(*xspi_cmd));
+	if (!xspi_cmd) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	/* Copy settings to device descriptor */
+	xspi_desc->device_id = param->device_id;
+	xspi_desc->max_speed_hz = param->max_speed_hz;
+	xspi_desc->extra = sdesc;
+	xspi_desc->platform_ops = param->platform_ops;
+
+	/* Assign the command pointer with the allocated memory */
+	sdesc->command = xspi_cmd;
+
+	/* Get the input clock for the XSPI */
+	if (sinit->get_input_clock) {
+		sdesc->input_clock = sinit->get_input_clock();
+	}
+
+	sdesc->sample_shifting = sinit->sample_shifting;
+	sdesc->cs_high_time = sinit->cs_high_time;
+	sdesc->fifo_threshold = sinit->fifo_threshold;
+
+	/* Initialize the XSPI peripheral */
+	ret = stm32_xspi_peripheral_init(xspi_desc, param);
+	if (ret)
+		goto error;
+
+	/* Configure the XSPI with command */
+	ret = stm32_xspi_configure(xspi_desc, param);
+	if (ret)
+		goto error;
+
+	/* If DMA for XSPI exists in init params, initialize DMA */
+	if (sinit->dma_init) {
+		ret = no_os_dma_init(&sdesc->dma_desc, sinit->dma_init);
+		if (ret)
+			goto error;
+
+		if (sinit->dma_ch) {
+			sdesc->dma_ch = &sdesc->dma_desc->channels[0];
+			sdesc->dma_ch->id = sinit->dma_ch->hdma;
+			sdesc->dma_ch->extra = sinit->dma_ch;
+			sdesc->dma_ch->irq_num = sinit->irq_num;
+		}
+	}
+
+	*desc = xspi_desc;
+
+	return 0;
+
+error:
+	no_os_dma_remove(sdesc->dma_desc);
```

Comment:
`sdesc` may not be allocated at this point if `xspi_desc` alloc fails. Also, please initialize all the pointers you're calling `free()` on here to `NULL`. It's undefined behavior to free a random pointer (which doesn't point to a memory region previously allocated).